src/Nethermind/Nethermind.TxPool/Collections/BlobTxDistinctSortedPool.cs

```diff
@@ -0,0 +1,70 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.Collections.Generic;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Logging;
+
+namespace Nethermind.TxPool.Collections;
+
+public class BlobTxDistinctSortedPool : TxDistinctSortedPool
+{
+    protected const int MaxNumberOfBlobsInBlock = (int)(Eip4844Constants.MaxBlobGasPerBlock / Eip4844Constants.BlobGasPerBlob);
+
+    public BlobTxDistinctSortedPool(int capacity, IComparer<Transaction> comparer, ILogManager logManager)
+        : base(capacity, comparer, logManager)
+    {
+    }
+
+    protected override IComparer<Transaction> GetReplacementComparer(IComparer<Transaction> comparer)
+        => comparer.GetBlobReplacementComparer();
+
+    public virtual bool TryInsertBlobTx(Transaction fullBlobTx, out Transaction? removed)
+        => base.TryInsert(fullBlobTx.Hash, fullBlobTx, out removed);
+
+    public virtual bool TryGetBlobTx(ValueKeccak hash, out Transaction? fullBlobTx)
+        => base.TryGetValue(hash, out fullBlobTx);
+
+    public virtual IEnumerable<Transaction> GetBlobTransactions()
+    {
+        int pickedBlobs = 0;
+        List<Transaction>? blobTxsToReadd = null;
+
+        while (pickedBlobs < MaxNumberOfBlobsInBlock)
+        {
+            Transaction? bestTx = GetFirsts().Min;
+
+            if (bestTx?.Hash is null || bestTx.BlobVersionedHashes is null)
+            {
+                break;
+            }
+
+            if (pickedBlobs + bestTx.BlobVersionedHashes.Length <= MaxNumberOfBlobsInBlock)
+            {
+                yield return bestTx;
+                pickedBlobs += bestTx.BlobVersionedHashes.Length;
+            }
+
+            if (TryRemove(bestTx.Hash))
+            {
+                blobTxsToReadd ??= new(MaxNumberOfBlobsInBlock);
+                blobTxsToReadd.Add(bestTx!);
+            }
+        }
```

Comment:
The problem with picking blobs in `TxPoolTxSource` is that `TxPoolTxSource` is allocating array of all txs. Because of that I decided to move picking to `PersistentTxDistinctSortedPool` and use light blob-tx equivalences for picking best txs and load full versions of only picked ones. And allocate in `TxPoolTxSource` only up to 6 blob txs.

I have another idea for it - to expose to `TxPoolTxSource` collection of light blob equivalences and load full tx only when is considered to be returned. So we will still load only up to 6 full txs from db, but we will additionally create a copy of light blob tx collection, so up to 16k txs. We are behaving that way with non-blob txs, but it is only 2k. Thanks to that we will be able to drop this logic from `PersistentTxDistinctSortedPool` and just use snapshot of light equivalences sorted by sender. I will try that approach and we can choose which one is the lesser evil

Reply:
It's refactored as described above - using light blob txs and loading full one only when picked to the block so max 6 txs per produced block